File: algorithm/力扣热题100/2_两数相加.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    const res = { next: null, val: 0 }
    let p = res, carry = 0
    while (l1 || l2 || carry) {
        if (l1) {
            carry += l1.val
            l1 = l1.next
        }
        if (l2) {
            carry += l2.val
            l2 = l2.next
        }
        p.next = {
            val: carry % 10,
            next: null
        }
        p = p.next
        carry = Math.floor(carry / 10)
    }
    return res.next
}

File: algorithm/力扣热题100/2_两数相加.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	carry := 0
	res := ListNode{
		Val:  0,
		Next: nil,
	}
	p := &res
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		p.Next = &ListNode{
			Val:  carry % 10,
			Next: nil,
		}
		p = p.Next
		carry /= 10
	}
	return res.Next
}
